File: howl-flutter/lib/pointer_input.dart
import 'dart:ui';

import 'package:flutter/gestures.dart';

import 'protocol.dart';

final class TerminalPointerLocation {
  const TerminalPointerLocation({
    required this.row,
    required this.column,
    required this.pixelX,
    required this.pixelY,
  });

  final int row;
  final int column;
  final int pixelX;
  final int pixelY;
}

final class TerminalPointerGeometry {
  const TerminalPointerGeometry({
    required this.viewport,
    required this.rows,
    required this.columns,
    required this.cellWidth,
    required this.rowHeight,
  });

  final Size viewport;
  final int rows;
  final int columns;
  final double cellWidth;
  final double rowHeight;

  TerminalPointerLocation? locate(Offset position) {
    if (!viewport.width.isFinite ||
        !viewport.height.isFinite ||
        rows <= 0 ||
        columns <= 0 ||
        cellWidth <= 0 ||
        rowHeight <= 0) {
      return null;
    }
    final terminalWidth = columns * cellWidth;
    final terminalHeight = rows * rowHeight;
    if (terminalWidth > viewport.width || terminalHeight > viewport.height) {
      return null;
    }
    final left = (viewport.width - terminalWidth) / 2;
    final top = (viewport.height - terminalHeight) / 2;
    final x = position.dx - left;
    final y = position.dy - top;
    if (x < 0 || y < 0 || x >= terminalWidth || y >= terminalHeight) {
      return null;
    }
    return TerminalPointerLocation(
      row: (y / rowHeight).floor(),
      column: (x / cellWidth).floor(),
      pixelX: x.floor(),
      pixelY: y.floor(),
    );
  }
}

final class TerminalPointerAdapter {
  final Map<int, _PointerState> _states = <int, _PointerState>{};

  List<HowlMouseInput> translate(
    PointerEvent event, {
    required TerminalPointerGeometry geometry,
    required int modifiers,
  }) {
    if (!_supportedKind(event.kind) || event is PointerSignalEvent) {
      return const <HowlMouseInput>[];
    }

    final previous = _states[event.pointer];
    final mapped = geometry.locate(event.localPosition);
    final location = mapped ?? previous?.location;
    final current = _howlButtons(event.buttons);
    final before = previous?.buttons ?? 0;

    if (event is PointerCancelEvent) {
      _states.remove(event.pointer);
      if (location == null) return const <HowlMouseInput>[];
      return _releaseTransitions(
        before,
        0,
        location: location,
        modifiers: modifiers,
      );
    }

    if (mapped != null || previous != null) {
      if (current == 0 && event is PointerUpEvent) {
        _states.remove(event.pointer);
      } else {
        _states[event.pointer] = _PointerState(
          buttons: current,
          location: mapped ?? previous!.location,
        );
      }
    }
    if (location == null) return const <HowlMouseInput>[];

    if (event is PointerDownEvent) {
      return _pressTransitions(
        before,
        current,
        location: location,
        modifiers: modifiers,
      );
    }
    if (event is PointerUpEvent) {
      return _releaseTransitions(
        before,
        current,
        location: location,
        modifiers: modifiers,
      );
    }
    if (event is PointerMoveEvent || event is PointerHoverEvent) {
      if (mapped == null) return const <HowlMouseInput>[];
      return <HowlMouseInput>[
        _input(
          kind: HowlWire.mouseMove,
          button: HowlWire.mouseNone,
          buttonsDown: current,
          location: location,
          modifiers: modifiers,
        ),
      ];
    }
    return const <HowlMouseInput>[];
  }

  List<HowlMouseInput> _pressTransitions(
    int before,
    int after, {
    required TerminalPointerLocation location,
    required int modifiers,
  }) {
    final changed = after & ~before;
    if (changed == 0) return const <HowlMouseInput>[];
    final output = <HowlMouseInput>[];
    var held = before;
    for (final bit in const <int>[1, 2, 4]) {
      if (changed & bit == 0) continue;
      held |= bit;
      output.add(
        _input(
          kind: HowlWire.mousePress,
          button: _buttonForBit(bit),
          buttonsDown: held,
          location: location,
          modifiers: modifiers,
        ),
      );
    }
    return output;
  }

  List<HowlMouseInput> _releaseTransitions(
    int before,
    int after, {
    required TerminalPointerLocation location,
    required int modifiers,
  }) {
    final changed = before & ~after;
    if (changed == 0) return const <HowlMouseInput>[];
    final output = <HowlMouseInput>[];
    var held = before;
    for (final bit in const <int>[1, 2, 4]) {
      if (changed & bit == 0) continue;
      held &= ~bit;
      output.add(
        _input(
          kind: HowlWire.mouseRelease,
          button: _buttonForBit(bit),
          buttonsDown: held,
          location: location,
          modifiers: modifiers,
        ),
      );
    }
    return output;
  }

  HowlMouseInput _input({
    required int kind,
    required int button,
    required int buttonsDown,
    required TerminalPointerLocation location,
    required int modifiers,
  }) => HowlMouseInput(
    kind: kind,
    button: button,
    modifiers: modifiers,
    buttonsDown: buttonsDown,
    row: location.row,
    column: location.column,
    pixelX: location.pixelX,
    pixelY: location.pixelY,
  );
}

final class _PointerState {
  const _PointerState({required this.buttons, required this.location});

  final int buttons;
  final TerminalPointerLocation location;
}

bool _supportedKind(PointerDeviceKind kind) =>
    kind == PointerDeviceKind.mouse ||
    kind == PointerDeviceKind.stylus ||
    kind == PointerDeviceKind.invertedStylus;

int _howlButtons(int flutterButtons) {
  var result = 0;
  if (flutterButtons & kPrimaryButton != 0) result |= 1;
  if (flutterButtons & kTertiaryButton != 0) result |= 2;
  if (flutterButtons & kSecondaryButton != 0) result |= 4;
  return result;
}

int _buttonForBit(int bit) => switch (bit) {
  1 => HowlWire.mouseLeft,
  2 => HowlWire.mouseMiddle,
  4 => HowlWire.mouseRight,
  _ => HowlWire.mouseNone,
};

File: howl-web/render/web/pointer_input.mjs
import {MouseButton, MouseKind} from './input.mjs';

export class TerminalPointerGeometry {
  constructor({left, top, width, height, surfaceWidth, surfaceHeight, cellWidth, cellHeight}) {
    Object.assign(this, {left, top, width, height, surfaceWidth, surfaceHeight, cellWidth, cellHeight});
  }

  locate({clientX, clientY}) {
    const {left, top, width, height, surfaceWidth, surfaceHeight, cellWidth, cellHeight} = this;
    if (![left, top, width, height, surfaceWidth, surfaceHeight, cellWidth, cellHeight].every(Number.isFinite) ||
        width <= 0 || height <= 0 || surfaceWidth <= 0 || surfaceHeight <= 0 || cellWidth <= 0 || cellHeight <= 0 ||
        surfaceWidth % cellWidth !== 0 || surfaceHeight % cellHeight !== 0) return null;
    const cssX = clientX - left, cssY = clientY - top;
    if (cssX < 0 || cssY < 0 || cssX >= width || cssY >= height) return null;
    const pixelX = Math.floor(cssX * surfaceWidth / width);
    const pixelY = Math.floor(cssY * surfaceHeight / height);
    if (pixelX < 0 || pixelY < 0 || pixelX >= surfaceWidth || pixelY >= surfaceHeight) return null;
    return {
      row:Math.floor(pixelY / cellHeight), column:Math.floor(pixelX / cellWidth),
      pixelX, pixelY,
    };
  }
}

export class TerminalPointerAdapter {
  constructor() { this.states = new Map(); }
  clear() { this.states.clear(); }

  translate(event, {geometry, modifiers = 0}) {
    if (!supportedPointerType(event.pointerType) || !geometry) return [];
    const pointerId = event.pointerId ?? 0;
    const previous = this.states.get(pointerId);
    const mapped = geometry.locate(event);
    const location = mapped ?? previous?.location;
    const current = howlButtons(event.buttons ?? 0);
    const before = previous?.buttons ?? 0;

    if (event.type === 'pointercancel') {
      this.states.delete(pointerId);
      return location == null ? [] : releaseTransitions(before, 0, location, modifiers);
    }

    if (mapped != null || previous != null) {
      if (current === 0 && event.type === 'pointerup') this.states.delete(pointerId);
      else this.states.set(pointerId, {buttons:current, location:mapped ?? previous.location});
    }
    if (location == null) return [];
    if (event.type === 'pointerdown') return pressTransitions(before, current, location, modifiers);
    if (event.type === 'pointerup') return releaseTransitions(before, current, location, modifiers);
    if (event.type === 'pointermove') return mapped == null ? [] : [mouseInput(MouseKind.move, MouseButton.none, current, location, modifiers)];
    return [];
  }

  wheel(event, {geometry, modifiers = 0}) {
    if (!geometry || !Number.isFinite(event.deltaY) || event.deltaY === 0) return null;
    const location = geometry.locate(event);
    if (location == null) return null;
    return mouseInput(
      MouseKind.wheel,
      event.deltaY < 0 ? MouseButton.wheelUp : MouseButton.wheelDown,
      0,
      location,
      modifiers,
    );
  }
}

// One mouse move may be on the wire while one newer move is retained locally.
// Button transitions stay outside this scheduler as ordered control barriers.
export class LatestPointerMoveScheduler {
  constructor({send, onError = () => {}}) {
    if (typeof send !== 'function') throw new Error('pointer move scheduler requires send');
    this.send = send; this.onError = onError; this.latest = null; this.running = false;
  }
  push(input) { this.latest = input; if (!this.running) void this.#drain(); }
  clear() { this.latest = null; }
  async #drain() {
    this.running = true;
    try {
      while (this.latest != null) {
        const input = this.latest; this.latest = null;
        await this.send(input);
      }
    } catch (error) {
      this.latest = null;
      this.onError(error);
    } finally {
      this.running = false;
      if (this.latest != null) void this.#drain();
    }
  }
}

function supportedPointerType(value) { return value === 'mouse' || value === 'pen'; }
function howlButtons(value) {
  let out = 0;
  if (value & 1) out |= 1; // DOM primary -> Howl left.
  if (value & 4) out |= 2; // DOM auxiliary -> Howl middle.
  if (value & 2) out |= 4; // DOM secondary -> Howl right.
  return out;
}
function buttonForBit(bit) { return bit === 1 ? MouseButton.left : bit === 2 ? MouseButton.middle : bit === 4 ? MouseButton.right : MouseButton.none; }
function mouseInput(kind, button, buttonsDown, location, modifiers) { return {kind, button, buttonsDown, modifiers, ...location}; }
function pressTransitions(before, after, location, modifiers) {
  const changed = after & ~before, out = []; let held = before;
  for (const bit of [1,2,4]) if (changed & bit) { held |= bit; out.push(mouseInput(MouseKind.press, buttonForBit(bit), held, location, modifiers)); }
  return out;
}
function releaseTransitions(before, after, location, modifiers) {
  const changed = before & ~after, out = []; let held = before;
  for (const bit of [1,2,4]) if (changed & bit) { held &= ~bit; out.push(mouseInput(MouseKind.release, buttonForBit(bit), held, location, modifiers)); }
  return out;
}
